src/master/master.test.ts

```diff
@@ -202,6 +202,72 @@ describe('update', () => {
     expect(sendAll).toHaveBeenCalled();
   });
 
+  test('allow execution of moves with ignoreStaleStateID truthy', async () => {
+    const game = {
+      moves: {
+        A: G => G,
+        B: {
+          move: G => G,
+          ignoreStaleStateID: true,
+        },
+      },
+    };
+
+    const send = jest.fn();
+    const master = new Master(game, new InMemory(), TransportAPI(send));
+
+    const actionA = ActionCreators.makeMove(
+      'A',
+      ['not ignore stale state'],
+      '0'
+    );
+    const actionB = ActionCreators.makeMove('B', ['ignore stale state'], '0');
+
+    // test: ping-pong two moves, then sync and check the log
+    await master.onSync('matchID', '0', 2);
+    await master.onUpdate(actionA, 0, 'matchID', '0');
+    await master.onUpdate(actionB, 0, 'matchID', '0');
```

Comment:
Well, I made that through unit tests and it didn't happened.  :smile:

Reply:
The test is using the synchronous `InMemory` storage, so I’m not sure it fully captures the scenario above. Maybe it would be possible with some wrapper around `InMemory` that adds some artificial delays. Something like:

```ts
class AsyncInMemory extends StorageAPI.Async {
  db: InMemory;

  constructor() {
    super();
    this.db = new InMemory();
  }

  private sleep(interval: number) {
    return new Promise(resolve => void setTimeout(resolve, interval));
  }

  async createMatch(id: string, opts: StorageAPI.CreateMatchOpts) {
    await this.sleep(100);
    return this.db.createMatch(id, opts);
  }

  // etc.
}
```

Reply:
Well, I've implemented the AsyncInMemory class and that strange behaviour you predicted happened: one player had it's action not executed or overwritten. I believe there's something regarding the order of execution of the promises. Strange that on my game, using Flatfile, it didn't happened.

```
class InMemoryAsync extends StorageAPI.Async {
  db: InMemory;

  constructor() {
    super();
    this.db = new InMemory();
  }

  async connect() {
    await this.sleep(100);
  }

  private sleep(interval: number): Promise<void> {
    return new Promise(resolve => void setTimeout(resolve, interval));
  }

  /**
   * @param id
   * @param opts
   * @override
   */
  async createMatch(id: string, opts: StorageAPI.CreateMatchOpts) {
    await this.sleep(100);
    this.db.createMatch(id, opts);
  }

  async setMetadata(matchID: string, metadata: Server.MatchData) {
    await this.sleep(100);
    this.db.setMetadata(matchID, metadata);
  }

  async setState(matchID: string, state: State, deltalog?: LogEntry[]) {
    await this.sleep(100);
    this.db.setState(matchID, state, deltalog)
  }

  async fetch<O extends StorageAPI.FetchOpts>(
    matchID: string,
    opts: O
  ): Promise<StorageAPI.FetchResult<O>> {
    await this.sleep(100);
    return this.db.fetch(matchID, opts);
  }

  async wipe(matchID: string) {
    await this.sleep(100);
    this.db.wipe(matchID);
  }

  /**
   * @param opts
   * @override
   */
  async listMatches(opts?: StorageAPI.ListMatchesOpts): Promise<string[]> {
    await this.sleep(100);
    return this.db.listMatches(opts);
  }
}
```

Reply:
Maybe we should implement a per-match queue system on the infrastructure level. I was thinking to use the PQueue package (https://www.npmjs.com/package/p-queue). What do you think?

Reply:
Thanks for testing this @evandroabukamel!

I actually thought about the `FlatFile` storage first (when I was suggesting to write a test with async storage), but remembered that someone added some concurrency management to `FlatFile` (#705). So I expect `FlatFile` would not produce this bug, but other asynchronous storage implementations could, because they won’t generally implement any queuing.

I guess a per-match queue would need to be added to [the server transport layer](https://github.com/boardgameio/boardgame.io/blob/9fcbed76d3fdd196b47e052d4ae2543e7d517a74/src/server/transport/socketio.ts#L121-L130).

Reply:
That's what I thought but it will not solve the case on the unit test since it is using the master directly. Wouldn't it be better to create a queue on the master.ts?

Reply:
Ok, @delucis, I implemented a per-match queue where you said, taking care to delete the queue when no more players are connected to a given match. Unfortunately I'm not sure how to write unit tests for that. :disappointed: 
I'm gonna try it on my game to see using InMemory database to see if it works.

Reply:
Oh, sorry, I missed this comment. Yes, testing this is pretty tricky. The socket tests are already pretty thin (there are no assertions in [the sync/update tests](https://github.com/boardgameio/boardgame.io/blob/d071e98c0868ea687d85168649e5d6302fb15229/src/server/transport/socketio.test.ts#L180-L194)) and I’m not immediately sure how to do improve them. I guess if we test the queue sufficiently, checking `onUpdate` is called sequentially, that should be enough. I can try and take a look when I have time to see if I can work out how this might be done.